src/transports/tcp/TcpDuplexConnection.cpp

```diff
@@ -127,76 +136,80 @@ class TcpReaderWriter : public ::folly::AsyncTransportWrapper::WriteCallback,
+  // callbacks even after DuplexConnection releases references to this
   std::shared_ptr<TcpReaderWriter> self_;
 };
 
 class TcpOutputSubscriber
-    : public SubscriberBaseT<std::unique_ptr<folly::IOBuf>> {
+    : public Subscriber<std::unique_ptr<folly::IOBuf>> {
  public:
-  TcpOutputSubscriber(
-      std::shared_ptr<TcpReaderWriter> tcpReaderWriter,
-      folly::Executor& executor)
-      : ExecutorBase(executor), tcpReaderWriter_(std::move(tcpReaderWriter)) {}
+  explicit TcpOutputSubscriber(
+      std::shared_ptr<TcpReaderWriter> tcpReaderWriter)
+      : tcpReaderWriter_(std::move(tcpReaderWriter)) {
+    CHECK(tcpReaderWriter_);
+  }
 
-  void onSubscribeImpl(
-      std::shared_ptr<Subscription> subscription) noexcept override {
+  void onSubscribe(
+      yarpl::Reference<Subscription> subscription) noexcept override {
+    CHECK(subscription);
     if (!tcpReaderWriter_) {
       LOG(ERROR) << "trying to resubscribe on a closed subscriber";
       subscription->cancel();
       return;
     }
-
     tcpReaderWriter_->setOutputSubscription(std::move(subscription));
   }
 
-  void onNextImpl(std::unique_ptr<folly::IOBuf> element) noexcept override {
+  void onNext(std::unique_ptr<folly::IOBuf> element) noexcept override {
     CHECK(tcpReaderWriter_);
     tcpReaderWriter_->send(std::move(element));
   }
 
-  void onCompleteImpl() noexcept override {
+  void onComplete() noexcept override {
     CHECK(tcpReaderWriter_);
-    auto tcpReaderWriter = std::move(tcpReaderWriter_);
-    tcpReaderWriter->close();
+    tcpReaderWriter_->setOutputSubscription(nullptr);
   }
 
-  void onErrorImpl(folly::exception_wrapper ew) noexcept override {
+  void onError(std::exception_ptr ex) noexcept override {
     CHECK(tcpReaderWriter_);
     auto tcpReaderWriter = std::move(tcpReaderWriter_);
-    tcpReaderWriter->closeErr(std::move(ew));
+    tcpReaderWriter->closeErr(folly::exception_wrapper(std::move(ex)));
   }
 
  private:
   std::shared_ptr<TcpReaderWriter> tcpReaderWriter_;
 };
 
-class TcpInputSubscription : public SubscriptionBase {
+class TcpInputSubscription : public Subscription {
  public:
-  TcpInputSubscription(
-      std::shared_ptr<TcpReaderWriter> tcpReaderWriter,
-      folly::Executor& executor)
-      : ExecutorBase(executor), tcpReaderWriter_(std::move(tcpReaderWriter)) {
+  explicit TcpInputSubscription(
+      std::shared_ptr<TcpReaderWriter> tcpReaderWriter)
+      : tcpReaderWriter_(std::move(tcpReaderWriter)) {
     CHECK(tcpReaderWriter_);
   }
 
-  void requestImpl(size_t) noexcept override {
-    // TcpDuplexConnection doesn't support proper flow control.
+  void request(int64_t n) noexcept override {
+    DCHECK(tcpReaderWriter_);
+    DCHECK(n == kMaxRequestN)
+        << "TcpDuplexConnection doesnt support proper flow control";
   }
 
-  void cancelImpl() noexcept override {
-    tcpReaderWriter_->close();
+  void cancel() noexcept override {
+    tcpReaderWriter_->setInput(nullptr);
```

Comment:
We would like to be able to call the setInput & getOutput functions on an instance of TcpReaderWriter multiple times. So when an instance of TcpInputSubscription calls cancel, it should not close the connection. One subscription/subscriber might cancel and then can be replaced with another one.